lib/recursive-methods.rb

```diff
@@ -1,49 +1,94 @@
+require "pry"
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
+# Space complexity: O(n) where the stack is n levels deep
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  if (n == 1) || (n == 0)
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
+# Space complexity: O(n) where n levels are required, each requiring 1 char less of memory
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 1 || s.length == 0
+    return s
+  else
+    return s[-1] + reverse(s[0...-1])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n/2 recursive calls are made
+# Space complexity: O(n) where the max depth of the function is n/2
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 1 || s.length == 0
+    return s
+  else
+    s[-1] + reverse_inplace(s[1...-1]) + s[0]
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n recursive calls are made
+# Space complexity: O(n) where the max depth is n
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n <= 0
+    return 0
+  else
+    return 2 + bunny(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n/2 recursive calls are made
+# Space complexity: O(n) where the max depth of the function is n/2
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return true
+  elsif s.length == 1
+    return false
+  elsif s[0] == s[-1]
+    return false
+  else
+    nested(s[1...s.length - 1])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because there are 2n-1 calls each time.
+# Space complexity: O(log n) because at most the algorithm goes log2 n levels deep
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  return false if array.length == 0
+  if array.length == 1
+    return true if array[0] == value
+  else
+    first_half = array[0...(array.length / 2)]
+    second_half = array[(array.length / 2)...array.length]
+    return true if search(first_half, value) || search(second_half, value)
+  end
+  return false
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n/2 recursive calls are made
```

Comment:
Not quite, as `s[1..s.length-1]` makes a **new** string with n-2 characters, and since you do this every recursive call... you end up with O(n * n) = O(n^2).

There is another way by using a helper method and keeping track of the current indexes which is O(n).

Reply:
Gotcha. For some reason I thought accessing a slice of an array/string would access part of what was already stored in memory.

One question though - why would creating new strings affect the time complexity (I get why they would affect space complexity. I'm not iterating through the string (I looked it up and in Ruby, .length is O(1) and that's the only iteration-like thing that's happening.